astriaorg/astria
# feat(sequencer)!: allow configuring base address prefix

## Summary
This patch allows configuring the base address prefix in sequencer's genesis file and enforces this prefix in all actions.

Similar to the native fee, the base prefix is set once during `init-chain` and is then available globally and never changed.

## Background
https://github.com/astriaorg/astria/pull/1124 changed astria addresses from opaque bytes to bech32m addresses, which themselves are encoded as strings on the wire. Where previously sequencer implicitly assumed that all addresses were `"astria"` prefixed bech32m addresses, this is now enforced.

## Changes
- Update sequencer `GenesisState` to use the canonical serde implementation of all `astria-core` types.
- Add a field `.address_prefixes.base` to the `GenesisState`.
- Ensure that all addresses in `GenesisState` are consistenst with the prefix.
- Inject the genesis base address prefix into sequencer's global state.
- Add a check for the global address prefix to all sequencer actions that carry a `astria.primitive.v1.core.Address` (this is done via `AppHandler::check_stateless`). These are specifically:
    - `TransferAction`
    - `SudoAddressChangeAction`
    - `BridgeLockAction`
    - `BridgeSudoChangeAction`
    - `BridgeUnlockAction`
    - `InitBridgeAccountAction`
    - `IbcRelayerChangeAction`
- Slightly change all storage keys that use addresses to be explicit about using bytes and provide snapshot tests (specifically in `accounts::state_ext`, `bridge::state_ext`, `ibc::state_ext`)
- Removed: unsigned transaction parameters no longer require a bech32m chain ID name
- Update composer, bridge-withdrawer to have a configurable address prefix:
    - `ASTRIA_BRIDGE_WITHDRAWER_SEQUENCER_ADDRESS_PREFIX`
    - `ASTRIA_COMPOSER_SEQUENCER_ADDRESS_PREFIX`
- Update all `astria-cli` commands to take a configurable prefix.

## Testing
+ Update all tests and ensure they still run.
+ Add tests for genesis construction.
+ Add genesis snapshot tests.
+ Add snapshots tests for IBC denoms (to accompany the genesis snapshot)

## Breaking Changelist
This change is breaking at the network level requiring sequencer re-genesis. Non-bech32m addresses in actions are now rejected, as well as addresses that have a prefix other then the one set at genesis. In addition, the storage key for all collections involving an address have changed.
